### PR TITLE
Allow any combination of query and reference sets

### PR DESCRIPTION
**Pull request by mlopezantequera**
This PR adds the capability to do any choice of datasets as the reference sets when evaluating.

Before, the choices were limited to these options:
- `compared_to_self`: each dataset split will refer to itself to find nearest neighbors.
- `compared_to_sets_combined`: each dataset split will refer to all provided splits to find nearest neighbors.
- `compared_to_training_set`: each dataset will refer to the training set to find nearest neighbors.

Now, in `splits_to_eval`, `test()` expects the names of the splits to use as query (as before), but also those to be used as reference, allowing full flexibility. For example, a dataset A can be used for training, and a validation set B can be evaluated using a dataset C as the reference.

There is still another use-case that might be interesting which is not yet supported: to evaluate the same query set on several different reference sets. I did not implement support for this to keep the PR light. A way to do it would be that all_accuracies uses a unique label for each combination of query and reference sets, instead of using only the query set as key. (It would also simplify the logic around logging, which now relies on an auxiliary variable "reference_split_names")

### CODE REVIEW

**Comment by KevinMusgrave** [comment]
This is great! I've merged the changes and they will be available soon on pypi as version 0.9.96.dev0.

> There is still another use-case that might be interesting which is not yet supported: to evaluate the same query set on several different reference sets

Good idea. I've created an issue (#252) so that it doesn't get forgotten.
